### PR TITLE
Update material text-field theme according to new MD specs

### PR DESCRIPTION
**Pull request by jouni**
Connects to #3

Update: this PR is no longer just about the text field styles. It contains the basis for the theme customizability and global style rules as well.

Not sure if this PR will need be split into multiple PRs in the end. I just this branch since I already had the text field styles ready here which allowed me to utilize the top level properties in it.

---
This change is [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-themes/2)

### CODE REVIEW

**Comment by jouni** [comment]
Still pending visual test updates.

**Comment by jouni** [comment]
Missing error message styles also.

File: demo/index.html [hunk 1]
```diff
@@ -7,6 +7,8 @@
 
   <title>vaadin-text-field</title>
 
+  <!-- Safari has severe rendering issues with native Shadow DOM -->
```
Comment (jouni) [inline demo/index.html]: https://bugs.webkit.org/show_bug.cgi?id=170762

Already fixed in WebKit. Not yet shipped, though.

**Comment by tomivirkki** [comment]
This is some epic stuff! :+1: 

---

Review status: 0 of 14 files reviewed at latest revision, 3 unresolved discussions, some commit checks broke.

---

*[demo/material/text-field.html, line 63 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-themes/2#-Kk_0gPxP7ZrwWDMdrQr:-Kk_0gPylxiVk_jopEdY:b24kv3o) ([raw file](https://github.com/vaadin/vaadin-themes/blob/7938b74f32498765dd10d4e97c86a3c1f5372eea/demo/material/text-field.html#L63)):*
> ```HTML
>     <demo-snippet>
>       <template>
>         <vaadin-text-field class="email" label="Email address" placeholder="Firstname" theme="always-float-label">
> ```

There's an experimental branch at https://github.com/vaadin/vaadin-text-field/tree/proto-input-bounds that applies the text-field some CSS properties to be used for placing the floating label when the prefix is present.

The set of required properties should be discussed (perhaps "--prefix-width" would work better than "--input-left")

---

*[material/grid.html, line 4 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-themes/2#-Kk_-CUgU9gBHRb4LEw3:-Kk_-CUgU9gBHRb4LEw4:b-q8cna5) ([raw file](https://github.com/vaadin/vaadin-themes/blob/7938b74f32498765dd10d4e97c86a3c1f5372eea/material/grid.html#L4)):*
> ```HTML
>   <style>
>     html {
>       --material-base-grid: 8px;
> ```

Should this property be adjustable? Other values than 8px don't seem to work right

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-themes/2#-:-Kk_-21U5YM902YlH4fF:b-eljav9)*

File: bower.json [hunk 2]
```diff
@@ -22,12 +22,12 @@
     "wct.conf.js"
   ],
   "dependencies": {
-    "polymer": "Polymer/polymer#^2.0.0-rc.1"
+    "polymer": "Polymer/polymer#^v2.0.0"
```
Comment (limonte) [inline bower.json]: What's the reason of adding `v`? There's no `v` in the `bower.json` standard: https://github.com/bower/spec/blob/master/json.md

File: demo/material/index.html [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        <div theme="secondary-text s-text">--material-font-size-xs</div>
+        <div theme="secondary-text s-text">theme="xs-text"</div>
+      </div>
+    </section>
+
+    <h3>Text emphasis</h3>
+    <p>The theme defines different font weights you can use. You can apply the weights either with CSS custom properties or with HTML attributes.</p>
+    <section class="sizes">
+      <div theme="background space-l">
+        <div theme="strong-text">Strong text</div>
```
Comment (limonte) [inline demo/material/index.html]: Strong text is not strong enough :) 



(Linux/Chrome)

File: material/colors.html [hunk 4]
```diff
@@ -0,0 +1,65 @@
+<custom-style>
+  <style>
+    html {
+      --material-body-text-color: var(--light-theme-text-color, rgba(0,0,0,0.87));
```
Comment (limonte) [inline material/colors.html]: CSS linter isn't set up yet, but use space after commas: `rgba(0,0,0,0.87) -> rgba(0, 0, 0, 0.87)`

File: demo/material/text-field.html [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    <demo-snippet>
+      <template>
+        <div theme="dark background space-m">
+          <!-- TODO: how to avoid the need to add the theme attribute here (as :host-context is no longer available)? -->
+
+          <vaadin-text-field theme="dark" label="Label" placeholder="Placeholder"></vaadin-text-field>
+
+          <vaadin-text-field theme="dark" label="Focused" value="Value" focused></vaadin-text-field>
+
+          <vaadin-text-field theme="dark" label="Numbers" value="Invalid" pattern="[0-9]+" invalid error-message="Numbers only!"></vaadin-text-field>
```
Comment (limonte) [inline demo/material/text-field.html]: Some weird glitching in Linux/Chrome:

recorded

File: material/typography.html [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    [theme~="strong-text"] {
+      font-weight: var(--material-font-weight-strong);
+    }
+
+    [theme~="light-text"] {
+      font-weight: var(--material-font-weight-light);
+    }
+
+    /* Remove margin from the first header element inside a padded container */
+    [theme*="space"] > h1:first-child,
```
Comment (limonte) [inline material/typography.html]: Are you sure that this block should be here and not as a standalone `reset.html` file?

**Comment by jouni** [comment]
Thanks for taking a look guys! This was mostly just my braindump at this point, not finalised code ready to be merged.

File: bower.json (hunk 2 above)
Comment (jouni) [inline bower.json]: No reason, I was just haphazardly trying to fix dependencies to be somewhat close to correct.

File: demo/material/index.html (hunk 3 above)
Comment (jouni) [inline demo/material/index.html]: Thanks, good point. What’s the actual font used on Linux? Apparently it doesn’t have a semibold version.

File: material/typography.html (hunk 6 above)
Comment (jouni) [inline material/typography.html]: Yeah, probably should be split into a separate file that the developer imports explicitly. I was just throwing stuff in here to see what sticks, and we can freely refactor this as we go along and learn more how we/users want to use these styles.

We’ve been talking about this with @tomivirkki today, and these will probably be moved into a separate style module.

File: material/typography.html [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+      --material-font-size-xxxl: calc(var(--material-font-size) * var(--material-font-scale-up) * var(--material-font-scale-up) * var(--material-font-scale-up) * var(--material-font-scale-up));
+      --material-font-size-xxl: calc(var(--material-font-size) * var(--material-font-scale-up) * var(--material-font-scale-up) * var(--material-font-scale-up));
+      --material-font-size-xl: calc(var(--material-font-size) * var(--material-font-scale-up) * var(--material-font-scale-up));
+      --material-font-size-l: calc(var(--material-font-size) * var(--material-font-scale-up));
+      --material-font-size-m: var(--material-font-size);
+      --material-font-size-s: calc(var(--material-font-size) / var(--material-font-scale-down));
+      --material-font-size-xs: calc(var(--material-font-size) / var(--material-font-scale-down) / var(--material-font-scale-down));
+
+      --material-font-weight-strong: 500;
```
Comment (limonte) [inline material/typography.html]: Related to strong font in Linux. Should this be `700`? https://css-tricks.com/almanac/properties/f/font-weight/

**Comment by samiheikki** [comment]
Reviewed 13 of 15 files at r3, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks broke.

---

*[demo/material/index.html, line 94 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-themes/2#-KkVtmP7or272heSAMMa-r3-94:-KkpOySKdDGpdg05qhDg:b-3cyr79) ([raw file](https://github.com/vaadin/vaadin-themes/blob/7938b74f32498765dd10d4e97c86a3c1f5372eea/demo/material/index.html#L94)):*
<details><summary><i>Previously, jouni (Jouni Koivuviita) wrote…</i></summary><blockquote>

Thanks, good point. What’s the actual font used on Linux? Apparently it doesn’t have a semibold version.
</blockquote></details>

Same with Windows/Chrome

---

*[demo/material/index.html, line 26 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-themes/2#-KkpNrgQu6qhmdWdM7F8:-KkpNrgREPdYyO9kCS2D:b-x8vrmm) ([raw file](https://github.com/vaadin/vaadin-themes/blob/b19604fdcbe2cfe99ce3569e3de6af90ecd5b940/demo/material/index.html#L26)):*
> ```HTML
>       display: flex;
>       flex-wrap: wrap;
>       align-items: center;
> ```

Font size variables are bit out of visible area (for example: `theme="xxxl-text"`) in small devices. Can we for example set `flex-direction: column;` on small (phones) devices.


---

*[demo/material/text-field.html, line 118 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-themes/2#-KkR2hS4YKoD9GH273SD-r3-118:-KkpPMZWuD8mJcgsO1j6:b-7dz53j) ([raw file](https://github.com/vaadin/vaadin-themes/blob/7938b74f32498765dd10d4e97c86a3c1f5372eea/demo/material/text-field.html#L118)):*
<details><summary><i>Previously, limonte (Limon Monte) wrote…</i></summary><blockquote>

The glitching in Linux/Chrome:

recorded

</blockquote></details>

Same with Windows/Chrome. Same with `focused` vaadin-text-field above.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-themes/2)*

**Pull request merged**